### PR TITLE
Allow interface functions to convert QNodes with pre-existing interfaces

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

The PennyLane interface functions `to_torch(qnode)`, `to_tf(qnode)`, and `to_autograd(qnode)` currently all make the assumption that the input QNode either has no attached interface (i.e., it is a 'bare' `JacobianQNode`) or that it has an autograd interface.

This causes issues if the QNode to be converted has the same interface as the one to be applied (see PL-451, whereby `to_tf(to_tf(qnode))` causes the gradient to be zeroed), or already has a pre-existing interface.

**Description of the Change:**

* All interfaces now keep a reference to the original bare QNode under the `_qnode` attribute.

* The interface functions return the input QNode, without modification, if the input QNode interface matches the function.

* If the input QNode has a pre-existing interface, then the interface functions instead return `to_interface(qnode._qnode)`

**Benefits:**

The interface functions contain logic detailing how to handle QNodes with no interfaces and pre-existing interfaces. Higher level abstractions that depend on these functions (such as the `qml.qnn` module) can now use these functions without worrying about the interface of input QNodes.

**Possible Drawbacks:**

While writing tests for `to_autograd()`, I noticed that it has a side-effect; it mutates the input QNode _in addition_ to returning a new QNode. I attempted to modify `to_autograd()` to perform a deep copy to ensure this no longer happens. This worked for simple QNodes, but QNode with tensor observables would not copy correctly --- the output would always be the ground state.

**Related GitHub Issues:** PL-451

### CODE REVIEW

**Comment by josh146** [comment]
@trbromley all tests seem to be passing, except for the Keras tests 🤔

**Comment by josh146** [comment]
> @trbromley all tests seem to be passing, except for the Keras tests 🤔

Fixed 💪 With `to_tf()`, care has to be taken to ensure that a new interface is provided for an existing TF QNode if the `dtype` argument changes.

**Comment by co9olguy** [comment]
@josh146 Is there any way to keep QNodes "interface-agnostic", and the interface is just a container/context which contains a "bare" QNode?

**Comment by josh146** [comment]
> @josh146 Is there any way to keep QNodes "interface-agnostic", and the interface is just a container/context which contains a "bare" QNode?

Maybe I'm not fully following, but this is (to some extent) how it is done currently with both the TF and Torch interfaces.

* The TF interface contains a 'bare' QNode internally, that is wrapped by the TF `custom_gradient` function. The bare QNode is stored as a private attribute.

* The Torch interface contains a 'bare' QNode internally. The 'container' in this case is the `torch.nn.Function` class. The bare QNode is stored as a private attribute.

It is difficult to standardize the interface containers much further, since TF requires that the container be `tf.custom_gradient`, and Torch requires that the container be `torch.nn.Function.apply`.

The one exception to this is the Autograd interface, where we actually monkeypatch the bare qnode, adding the `AutogradQNode` class mixin which registers the `.jacobian()` method as an autograd primitive bound to `__call__()`. This is peculiar to Autograd --- we originally attempted to apply the same approach as for TF and Torch (container class which contains the custom gradient logic), but couldn't get this to work. Autograd seems to disallow splitting the autograd primitive and the function/method it is bound to between different objects.

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -2,26 +2,26 @@
...
 
-* Adds a device test suite, located at `pennylane/plugins/tests`, which can be used 
-  to run generic tests on core or external devices calling 
-  
-  >>> pytest pennylane/plugins/tests --device default.qubit --shots 1234 --analytic False                                                                                                                                                                                                                                                                    
-  
+* Adds a device test suite, located at `pennylane/plugins/tests`, which can be used
+  to run generic tests on core or external devices calling
+
+  >>> pytest pennylane/plugins/tests --device default.qubit --shots 1234 --analytic False
```
Comment (trbromley) [inline .github/CHANGELOG.md:8]: Might be good to surround with `pycon` as it's currently rendering as:
image

Sorry I couldn't make a direct suggestion as the triple backticks don't work with github suggestions.

File: pennylane/interfaces/autograd.py [hunk 2]
```diff
@@ -99,5 +106,6 @@ def gradient_product(g):
 
     # define the vector-Jacobian product function for AutogradQNode.evaluate
     autograd.extend.defvjp(AutogradQNode.evaluate, AutogradQNode.QNode_vjp, argnums=[1])
+    qnode._qnode = qnode  # pylint: disable=protected-access
```
Comment (trbromley) [inline pennylane/interfaces/autograd.py:109]: Could you explain this line? Is this for the case where `qnode_interface is None`?

File: pennylane/interfaces/tf.py [hunk 3]
```diff
@@ -76,6 +76,16 @@ def to_tf(qnode, dtype=None):
...
+    qnode_interface = getattr(qnode, "interface", None)
+
+    if qnode_interface == "tf":
+        if dtype == qnode._dtype:  # pylint: disable=protected-access
+            return qnode
+
+        qnode = qnode._qnode  # pylint: disable=protected-access
+
+    elif qnode_interface is not None:
+        qnode = qnode._qnode  # pylint: disable=protected-access
```
Comment (trbromley) [inline pennylane/interfaces/tf.py:88]:
```suggestion
    if qnode_interface == "tf" and dtype == qnode._dtype:
            return qnode
    elif qnode_interface is not None:
        qnode = qnode._qnode  # pylint: disable=protected-access
```
Would this work?

File: tests/interfaces/test_tf.py [hunk 4]
```diff
@@ -1270,3 +1270,76 @@ def circuit(x, y, z):
...
+
+        with tf.GradientTape() as tape:
+            res = converted_qnode(x)
+
+        assert np.allclose(res, tf.cos(x), atol=tol, rtol=0)
+
+        grad = tape.gradient(res, x)
+        assert np.allclose(grad, -tf.sin(x), atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("interface", [None, "autograd"])
```
Comment (trbromley) [inline tests/interfaces/test_tf.py:1329]: What is the motivation to split up these tests? For better monitoring if a test fails?

File: pennylane/interfaces/autograd.py (hunk 2 above)
Comment (trbromley) [reply pennylane/interfaces/autograd.py:109]: Ah is this part linked to the "possible drawbacks":
> While writing tests for to_autograd(), I noticed that it has a side-effect; it mutates the input QNode in addition to returning a new QNode. I attempted to modify to_autograd() to perform a deep copy to ensure this no longer happens. This worked for simple QNodes, but QNode with tensor observables would not copy correctly --- the output would always be the ground state.

File: .github/CHANGELOG.md (hunk 1 above)
Comment (josh146) [reply .github/CHANGELOG.md:8]: Oh! This came from a previous PR, but happy to change it. Not sure why it's showing up in this PR :/

File: tests/interfaces/test_tf.py (hunk 4 above)
Comment (josh146) [reply tests/interfaces/test_tf.py:1329]: Oh good catch, I'll merge them

File: pennylane/interfaces/autograd.py (hunk 2 above)
Comment (josh146) [reply pennylane/interfaces/autograd.py:109]: Ah, this is more because I want to be consistent with the other interfaces -- that a reference to the 'bare' QNode exists at `self._qnode`.

File: tests/interfaces/test_torch.py [hunk 5]
```diff
@@ -1216,3 +1216,70 @@ def circuit(x, y, z):
...
+        assert converted_qnode is not qnode
+        assert converted_qnode._qnode is getattr(qnode, "_qnode", qnode)
+
+        x_val = 0.4
+        x = torch.tensor(x_val, requires_grad=True)
+        res = converted_qnode(x)
+        res.backward()
+
+        assert np.allclose(res.detach().numpy(), np.cos(x_val), atol=tol, rtol=0)
+        assert np.allclose(x.grad, -np.sin(x_val), atol=tol, rtol=0)
```
Comment (trbromley) [inline tests/interfaces/test_torch.py:1285]:
```suggestion
    @pytest.mark.parametrize("interface", [None, "autograd", "tf"])
    def test_other_conversion(self, qnode, tol):
        """Tests that the to_torch() function correctly converts both tf and autograd qnodes and
        QNodes with no interface."""
        converted_qnode = to_torch(qnode)
        assert converted_qnode is not qnode
        assert converted_qnode._qnode is getattr(qnode, "_qnode", qnode)

        x_val = 0.4
        x = torch.tensor(x_val, requires_grad=True)
        res = converted_qnode(x)
        res.backward()

        assert np.allclose(res.detach().numpy(), np.cos(x_val), atol=tol, rtol=0)
        assert np.allclose(x.grad, -np.sin(x_val), atol=tol, rtol=0)
```
